core/src/apps/cardano/layout.py

```diff
@@ -265,43 +266,76 @@ async def confirm_datum_hash(datum_hash: bytes) -> None:
 
 
 async def confirm_inline_datum(first_chunk: bytes, inline_datum_size: int) -> None:
-    await _confirm_data_chunk(
+    await _confirm_tx_data_chunk(
         "confirm_inline_datum",
         "Inline datum",
         first_chunk,
         inline_datum_size,
     )
 
 
+async def confirm_message_payload(
+    payload_first_chunk: bytes,
+    payload_hash: bytes,
+    payload_size: int,
+    is_signing_hash: bool,
+    prefer_hex_display: bool = False,
+) -> None:
+    props: list[PropertyType]
+
+    if is_unambiguous_ascii(payload_first_chunk) and not prefer_hex_display:
+        props = _get_data_chunk_props(
+            "Message text", payload_first_chunk.decode("ascii"), payload_size
+        )
+    else:
+        props = _get_data_chunk_props("Message hex", payload_first_chunk, payload_size)
+
+    props.append(("Message hash:", payload_hash))
+
+    await confirm_properties(
+        "confirm_message_payload",
+        title="Confirm message hash" if is_signing_hash else "Confirm message",
+        props=props,
+        br_code=BRT_Other,
+    )
+
+
 async def confirm_reference_script(
     first_chunk: bytes, reference_script_size: int
 ) -> None:
-    await _confirm_data_chunk(
+    await _confirm_tx_data_chunk(
         "confirm_reference_script",
         "Reference script",
         first_chunk,
         reference_script_size,
     )
 
 
-async def _confirm_data_chunk(
-    br_type: str, title: str, first_chunk: bytes, data_size: int
-) -> None:
+def _get_data_chunk_props(
+    title: str, first_chunk: bytes | str, data_size: int
+) -> list[PropertyType]:
     MAX_DISPLAYED_SIZE = 56
     displayed_bytes = first_chunk[:MAX_DISPLAYED_SIZE]
     bytes_optional_plural = "byte" if data_size == 1 else "bytes"
-    props: list[tuple[str, bytes | None]] = [
+    props: list[PropertyType] = [
         (
             f"{title} ({data_size} {bytes_optional_plural}):",
             displayed_bytes,
         )
     ]
     if data_size > MAX_DISPLAYED_SIZE:
         props.append(("...", None))
```

Comment:
Hmm, I think it would be best if we showed the whole ASCII message even if it is long. 
For bytes in hex, showing the whole of the message does not really add security for most users --- they are likely to just be annoyed and skip it quickly. And the more paranoid users can avoid the feature (and ask us to show everything --- I doubt there will be demand for that, but we will see) or only use it when long messages are hashed.